### DISCUSSION TITLE
Please help us fix the NPC desync issue!

### DISCUSSION THREAD

**Discussion by hakusaro** [Pryaxis/TShock#1787] [Q&A]
We're looking for any information you have about the "NPC desync issue." This manifests in several ways, including

* Boss NPCs glitching around on screen
* Boss NPCs teleporting around the map
* Normal mobs teleporting and glitching out

If you experience anything weird with NPCs, please
* Get your players to screenshot the game during that time with the F8 network debug open
* Please also get a screenshot of task manager showing the graphs of logical processors.
  * Get the CPU info on *players* and *servers* if possible.
* Get the number of players.
* Get the difficulty.
* Get the team state.
* Get the CPU make/model and the RAM capacity.

We're trying to narrow down what causes it. We think:
* It might be difficulty related (Classic is okay, but Expert and Master may not be).
* It might be related to being on a team (not being on a team is okay, but being on a team is not).

We are almost certain that it is not **network capacity or throughput related.**

Please tell me those two things as well. Thanks!

**Comment by lflare** [comment]
Difficulty: Expert
Version: VANILLA 1.4.0.2
OS: Pterodactyl/Linux
RAM: 2GB
CPU: i7-8559U Virtualised 4 Cores
Network: Delicious Fiber

Mobs teleportation might have something to do with player count. The wall of flesh was actually possible to kill at 4 players online, but once we got 5 to 7 players, the wall of flesh became the god of death, teleporting past us faster than our mounts and instagibbing all of us.

It might ALSO be related to mob speed, since the teleportation is not as pronounced for slower mobs, but for faster mobs like the wall of flesh and queen bee, it becomes a game of thrones.

**Reply by hakusaro** [reply]
What's the processor on the server, and the client?

**Reply by lflare** [reply]
Server is running virtualised cores, since I have it running on my homelab. It ain't no Xeon, but it certainly isn't an Atom.

Clientwise, all of us are running varying rigs, with myself on a Ryzen 5 3600. All of us are experiencing the teleportation of damnation with the mobs.

**Reply by hakusaro** [reply]
Does anyone see at least one CPU thread pegged at 100% for a long period of time?

**Reply by lflare** [reply]
CPU usage on the server was definitely high, hovering around 100-200% (1-2 cores). No one in my stack was monitoring our own process threads at that point in time, since we figured it was a server issue.

**Reply by hakusaro** [reply]
We think it could be either, but we're not sure. It could honestly be both. Thank you for your report!

**Comment by matecopter** [comment]
- Difficulty: Master

- Server:
   RAM: 16GB
   CPU: AMD Opteron(tm) X3418 APU 
   OS:
    - Host: Linux/Fedora 31
    - Container (Docker): ryshe/terraria:1.4.0.2-pre2 (alpine 3.11.6)
    

- Client:
    OS: Solus/Linux
    RAM: 16GB
    CPU: Intel(R) Xeon(R) CPU E3-1231 v3 @ 3.40GHz
    Team: none

- Network: LAN

1. Tried to catch critters. As i was about to catch a bunny it vanished and reappeared once i moved away. 
![terraria_critter_desync](https://user-images.githubusercontent.com/37692639/82298505-843d6300-99b4-11ea-89a3-d45c48b8e5ef.png)
Client:
![terraria_critter_desync_client_cpu_usage](https://user-images.githubusercontent.com/37692639/82298616-acc55d00-99b4-11ea-98e3-1a7f7d52a72e.png)
Server:
![terraria_critter_desync_server_cpu_usage](https://user-images.githubusercontent.com/37692639/82298628-af27b700-99b4-11ea-90cb-0929aaaee86a.png)
Please note that there are a myriad of other containers running on my server, thus the CPU-Usage my not be that accurate.


2. Tried to open a chest and was not able to. Seemed to be out of sync. To save the position i set a warp point, then logged out and back in. When i warped to my previously set warp point i was warped to my home instead.
![terarria_chest_desync](https://user-images.githubusercontent.com/37692639/82298432-6839c180-99b4-11ea-8634-766eccd4e902.png)

**Reply by matecopter** [reply]
It seems that server and client get out of sync after a player death as all my desyncs have occurred after my character died and respawned.
![terraria_desync_after_death](https://user-images.githubusercontent.com/37692639/82301466-97523200-99b8-11ea-9a6a-ba2b9b192966.png)

**Reply by fernandofig** [reply]
I'm sorry I can't provide hard evidence to that effect at the moment, but I suppose it may be useful to report that my users observed exactly that behavior as well (desync issues after someone dies - they constantly have to leave and rejoin the server to temporarily workaround that).

**Comment by lflare** [comment]
Terraria 1.4.0.3 mostly mitigates this issue with client-side smoothing for NPCs.

**Comment by FraxinusErit** [comment]
Is this a current issue, or is this a purely CPU bound problem?
